### ISSUE TITLE
On all_gather in distributed training

### ISSUE THREAD

**Issue by seanzhuh** [3dlg-hcvc/DuoduoCLIP#10]
Hello, I want to do distributed training via [this line of code](https://github.com/3dlg-hcvc/DuoduoCLIP/blob/main/src/model/duoduoclip.py#L138), I notice that your pytorch version is 2.2.1 and I searched over pytorch documentation, it turns out there is no dist.nn.all_gather, only dist.all_gather, so when I replace dist.nn.all_gather with dist.all_gather, it shows no gradient, the all_gather operation won't keep track of gradient calculation, leading to no gradient is calculated. This guess comes from when I set trainer.device = 1, the loss decreases decently, but when trainer.device = 4 and using dist.all_gather(), the loss won't go down and higher than single gpu training. 

Here is how I do this
```
if self.hparams.devices > 1:
        text_feat_all = [torch.zeros_like(text_feat) for _ in range(self.hparams.devices)]
        image_feat_all = [torch.zeros_like(image_feat) for _ in range(self.hparams.devices)]
        pts_feat_all = [torch.zeros_like(pts_feat) for _ in range(self.hparams.devices)]
        pts_text_mask_all = [torch.zeros_like(pts_text_mask) for _ in range(self.hparams.devices)]
        dist.all_gather(text_feat_all, text_feat)
        dist.all_gather(image_feat_all, image_feat)
        dist.all_gather(pts_feat_all, pts_feat)
        dist.all_gather(pts_text_mask_all, pts_text_mask)
        text_feat = torch.cat(text_feat_all, dim=0)
        image_feat = torch.cat(image_feat_all, dim=0)
        pts_feat = torch.cat(pts_feat_all, dim=0)
        pts_text_mask = torch.cat(pts_text_mask_all, dim=0)
 
loss_dict = self.loss(pts_feat, image_feat, text_feat, pts_text_mask)
```

**Comment by hanhung** [comment]
Hello,

We referred to the usage of
```
torch.distributed.nn.all_gather
```
from the [openshape source code](https://github.com/Colin97/OpenShape_code/blob/master/src/train.py#L64) and [OpenCLIP](https://github.com/Colin97/OpenShape_code/blob/master/src/train.py#L64) implementations. However, I'm not sure yet why the function doesn't work anymore. Maybe it has been deprecated.

We will look into it and see if we can change to using another function.

**Comment by seanzhuh** [comment]
Thanks for your reply. It turns out we can't use
```
import torch
torch.distributed.nn.all_gather()
```
instead we should do
`from torch.distributed.nn import all_gather()`
